precompiles/conviction-voting/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+mod tests;
+
+type BalanceOf<Runtime> = <<Runtime as pallet_conviction_voting::Config>::Currency as Currency<
+	<Runtime as frame_system::Config>::AccountId,
+>>::Balance;
+type IndexOf<Runtime> = <<Runtime as pallet_conviction_voting::Config>::Polls as Polling<
+	Tally<
+		<<Runtime as pallet_conviction_voting::Config>::Currency as Currency<
+			<Runtime as frame_system::Config>::AccountId,
+		>>::Balance,
+		<Runtime as pallet_conviction_voting::Config>::MaxTurnout,
+	>,
+>>::Index;
+type ClassOf<Runtime> = <<Runtime as pallet_conviction_voting::Config>::Polls as Polling<
+	Tally<
+		<<Runtime as pallet_conviction_voting::Config>::Currency as Currency<
+			<Runtime as frame_system::Config>::AccountId,
+		>>::Balance,
+		<Runtime as pallet_conviction_voting::Config>::MaxTurnout,
+	>,
+>>::Class;
+
+/// A precompile to wrap the functionality from pallet-conviction-voting.
+pub struct ConvictionVotingPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+impl<Runtime> ConvictionVotingPrecompile<Runtime>
+where
+	Runtime: pallet_conviction_voting::Config + pallet_evm::Config + frame_system::Config,
+	BalanceOf<Runtime>: TryFrom<U256>,
+	<Runtime as frame_system::Config>::RuntimeCall:
+		Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<<Runtime as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as frame_system::Config>::RuntimeCall: From<ConvictionVotingCall<Runtime>>,
+	IndexOf<Runtime>: TryFrom<u32>,
+	ClassOf<Runtime>: TryFrom<u16>,
+{
+	/// Vote yes in a poll.
+	///
+	/// Parameters:
+	/// * poll_index: Index of poll
+	/// * vote_amount: Balance locked for vote
+	/// * conviction: Conviction multiplier for length of vote lock
+	#[precompile::public("voteYes(uint32,uint256,uint8)")]
+	fn vote_yes(
+		handle: &mut impl PrecompileHandle,
+		poll_index: u32,
+		vote_amount: U256,
+		conviction: u8,
+	) -> EvmResult {
+		let poll_index = Self::u32_to_index(poll_index).in_field("pollIndex")?;
+		let vote_amount = Self::u256_to_amount(vote_amount).in_field("voteAmount")?;
+		let conviction = Self::u8_to_conviction(conviction).in_field("conviction")?;
+
+		let vote = AccountVote::Standard {
+			vote: Vote {
+				aye: true,
+				conviction,
+			},
+			balance: vote_amount,
+		};
+
+		log::trace!(target: "conviction-voting-precompile",
+			"Voting {:?} on poll {:?}, with conviction {:?}",
+			true, poll_index, conviction
+		);
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		let call = ConvictionVotingCall::<Runtime>::vote { poll_index, vote }.into();
+
+		<RuntimeHelper<Runtime>>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		Ok(())
+	}
+
+	/// Vote no in a poll.
+	///
+	/// Parameters:
+	/// * poll_index: Index of poll
+	/// * vote_amount: Balance locked for vote
+	/// * conviction: Conviction multiplier for length of vote lock
+	#[precompile::public("voteNo(uint32,uint256,uint8)")]
+	fn vote_no(
```

Comment:
nit: this is very repetitive with `vote_yes`. I'm not sure why we made this two fns at the solidity interface level, but they don't have to be two fns here (less code duplication).

Reply:
I think having two separate functions is fine (a boolean is less of an intuitive thing IMHO). But what I do agree with is that the code should be merged for both

Reply:
We need to anticipate for the future vote with "abstain" making vote support 3 possibilities.

Reply:
I agree with @girazoki, the best is to create a `inner_vote` function that takes as parameter the nature of the vote in order to factorize the code.

Reply:
> We need to anticipate for the future vote with "abstain" making vote support 3 possibilities.

The backwards compatible way is to add a new `voteAbstain` function when this is exposed.

> I think having two separate functions is fine (a boolean is less of an intuitive thing IMHO). But what I do agree with is that the code should be merged for both

Agreed, doing this now